flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/LegacyScheduler.java

```diff
@@ -0,0 +1,639 @@
+			if (executionGraph.getState() == JobStatus.RUNNING) {
+				log.error(errorMessage, jobGraph.getJobID());
+			} else {
+				log.debug(errorMessage, jobGraph.getJobID());
+			}
+		}
+	}
+
+	@Override
+	public void declineCheckpoint(final DeclineCheckpoint decline) {
+		mainThreadExecutor.assertRunningInMainThread();
+
+		final CheckpointCoordinator checkpointCoordinator = executionGraph.getCheckpointCoordinator();
+
+		if (checkpointCoordinator != null) {
+			ioExecutor.execute(() -> {
+				try {
+					checkpointCoordinator.receiveDeclineMessage(decline);
+				} catch (Exception e) {
+					log.error("Error in CheckpointCoordinator while processing {}", decline, e);
+				}
+			});
+		} else {
+			String errorMessage = "Received DeclineCheckpoint message for job {} with no CheckpointCoordinator";
+			if (executionGraph.getState() == JobStatus.RUNNING) {
+				log.error(errorMessage, jobGraph.getJobID());
+			} else {
+				log.debug(errorMessage, jobGraph.getJobID());
+			}
+		}
+	}
+
+	@Override
+	public CompletableFuture<String> stopWithSavepoint(final String targetDirectory, final boolean advanceToEndOfEventTime) {
+		mainThreadExecutor.assertRunningInMainThread();
+
+		final CheckpointCoordinator checkpointCoordinator = executionGraph.getCheckpointCoordinator();
+
+		if (checkpointCoordinator == null) {
+			return FutureUtils.completedExceptionally(new IllegalStateException(
+				String.format("Job %s is not a streaming job.", jobGraph.getJobID())));
+		}
+
+		if (targetDirectory == null && !checkpointCoordinator.getCheckpointStorage().hasDefaultSavepointLocation()) {
+			log.info("Trying to cancel job {} with savepoint, but no savepoint directory configured.", jobGraph.getJobID());
+
+			return FutureUtils.completedExceptionally(new IllegalStateException(
+				"No savepoint directory configured. You can either specify a directory " +
+					"while cancelling via -s :targetDirectory or configure a cluster-wide " +
+					"default via key '" + CheckpointingOptions.SAVEPOINT_DIRECTORY.key() + "'."));
+		}
+
+		final long now = System.currentTimeMillis();
+
+		// we stop the checkpoint coordinator so that we are guaranteed
+		// to have only the data of the synchronous savepoint committed.
+		// in case of failure, and if the job restarts, the coordinator
+		// will be restarted by the CheckpointCoordinatorDeActivator.
+		checkpointCoordinator.stopCheckpointScheduler();
+
+		final CompletableFuture<String> savepointFuture = checkpointCoordinator
+			.triggerSynchronousSavepoint(now, advanceToEndOfEventTime, targetDirectory)
+			.handleAsync((completedCheckpoint, throwable) -> {
+				if (throwable != null) {
+					log.info("Failed during stopping job {} with a savepoint. Reason: {}", jobGraph.getJobID(), throwable.getMessage());
+					throw new CompletionException(throwable);
+				}
+				return completedCheckpoint.getExternalPointer();
+			}, mainThreadExecutor);
```

Comment:
Is the main thread executor really needed here? We don't modify state inside `handleAsync()`.

Reply:
I don't think so. I guess it should simply be `handle`